File: frontend/src/types/load.ts
export type PacketReadiness = {
  readiness_state?: string | null;
  ready_for_invoice?: boolean | null;
  ready_to_submit?: boolean | null;
  present_documents?: string[] | null;
  missing_required_documents?: {
    invoice?: string[] | null;
    submission?: string[] | null;
  } | null;
  blockers?: string[] | null;
  notes?: string[] | null;
};

export type LoadStatus =
  | "booked"
  | "in_transit"
  | "delivered"
  | "docs_received"
  | "docs_needs_attention"
  | "invoice_ready"
  | "submitted_to_broker"
  | "submitted_to_factoring"
  | "packet_rejected"
  | "resubmission_needed"
  | "advance_paid"
  | "reserve_pending"
  | "fully_paid"
  | "short_paid"
  | "disputed"
  | "archived";

export type LoadProcessingStatus =
  | "pending"
  | "in_progress"
  | "completed"
  | "failed";

export type Load = {
  id: string;
  load_number: string;
  status: LoadStatus;

  processing_status?: LoadProcessingStatus | null;

  customer_account_id?: string | null;
  driver_id?: string | null;
  broker_id?: string | null;

  rate_confirmation_number?: string | null;
  bol_number?: string | null;
  invoice_number?: string | null;

  broker_name_raw?: string | null;
  broker_email_raw?: string | null;

  pickup_date?: string | null;
  delivery_date?: string | null;

  pickup_location?: string | null;
  delivery_location?: string | null;

  gross_amount?: number | null;
  currency_code?: string | null;

  documents_complete?: boolean | null;
  packet_readiness?: PacketReadiness | null;
  has_ratecon?: boolean | null;
  has_bol?: boolean | null;
  has_invoice?: boolean | null;

  extraction_confidence_avg?: number | null;

  last_reviewed_by?: string | null;
  last_reviewed_at?: string | null;

  submitted_at?: string | null;
  funded_at?: string | null;
  paid_at?: string | null;

  notes?: string | null;

  created_at?: string | null;
  updated_at?: string | null;
};

File: backend/app/schemas/load.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.domain.enums.channel import Channel
from app.domain.enums.load_status import LoadStatus
from app.domain.enums.processing_status import ProcessingStatus


class ApiError(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str
    message: str
    details: dict[str, Any] | None = None


class LoadCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    customer_account_id: str
    driver_id: str
    broker_id: str | None = None
    source_channel: Channel = Channel.MANUAL
    load_number: str | None = Field(default=None, max_length=100)
    rate_confirmation_number: str | None = Field(default=None, max_length=100)
    bol_number: str | None = Field(default=None, max_length=100)
    invoice_number: str | None = Field(default=None, max_length=100)
    broker_name_raw: str | None = Field(default=None, max_length=255)
    broker_email_raw: str | None = Field(default=None, max_length=255)
    pickup_date: date | None = None
    delivery_date: date | None = None
    pickup_location: str | None = Field(default=None, max_length=255)
    delivery_location: str | None = Field(default=None, max_length=255)
    gross_amount: Decimal | None = Field(default=None, ge=Decimal("0"))
    currency_code: str = Field(default="USD", min_length=3, max_length=3)
    notes: str | None = None


class LoadUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    customer_account_id: str | None = None
    driver_id: str | None = None
    broker_id: str | None = None
    source_channel: Channel | None = None
    load_number: str | None = Field(default=None, max_length=100)
    rate_confirmation_number: str | None = Field(default=None, max_length=100)
    bol_number: str | None = Field(default=None, max_length=100)
    invoice_number: str | None = Field(default=None, max_length=100)
    broker_name_raw: str | None = Field(default=None, max_length=255)
    broker_email_raw: str | None = Field(default=None, max_length=255)
    pickup_date: date | None = None
    delivery_date: date | None = None
    pickup_location: str | None = Field(default=None, max_length=255)
    delivery_location: str | None = Field(default=None, max_length=255)
    gross_amount: Decimal | None = Field(default=None, ge=Decimal("0"))
    currency_code: str | None = Field(default=None, min_length=3, max_length=3)
    documents_complete: bool | None = None
    has_ratecon: bool | None = None
    has_bol: bool | None = None
    has_invoice: bool | None = None
    follow_up_required: bool | None = None
    notes: str | None = None


class LoadStatusUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    new_status: LoadStatus
    notes: str | None = None


class LoadRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: str
    organization_id: str
    customer_account_id: str
    driver_id: str
    broker_id: str | None = None
    source_channel: Channel
    status: LoadStatus
    processing_status: ProcessingStatus
    load_number: str | None = None
    rate_confirmation_number: str | None = None
    bol_number: str | None = None
    invoice_number: str | None = None
    broker_name_raw: str | None = None
    broker_email_raw: str | None = None
    pickup_date: date | None = None
    delivery_date: date | None = None
    pickup_location: str | None = None
    delivery_location: str | None = None
    gross_amount: Decimal | None = None
    currency_code: str
    documents_complete: bool
    has_ratecon: bool
    has_bol: bool
    has_invoice: bool
    extraction_confidence_avg: Decimal | None = None
    last_reviewed_by: str | None = None
    last_reviewed_at: datetime | None = None
    last_contacted_at: datetime | None = None
    follow_up_required: bool = False
    submitted_at: datetime | None = None
    funded_at: datetime | None = None
    paid_at: datetime | None = None
    notes: str | None = None
    created_at: datetime
    updated_at: datetime


class LoadListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: str
    customer_account_id: str
    driver_id: str
    broker_id: str | None = None
    source_channel: Channel
    status: LoadStatus
    processing_status: ProcessingStatus
    load_number: str | None = None
    rate_confirmation_number: str | None = None
    bol_number: str | None = None
    gross_amount: Decimal | None = None
    currency_code: str
    pickup_date: date | None = None
    delivery_date: date | None = None
    documents_complete: bool
    created_at: datetime
    updated_at: datetime


class WorkflowEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: str
    load_id: str
    event_type: str
    old_status: str | None = None
    new_status: str | None = None
    actor_staff_user_id: str | None = None
    actor_type: str
    created_at: datetime


class LoadResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    data: LoadRead
    meta: dict[str, Any] = Field(default_factory=dict)
    error: ApiError | None = None


class LoadListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    data: list[LoadListItem]
    meta: dict[str, Any]
    error: ApiError | None = None


class LoadTimelineResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    data: list[WorkflowEventRead]
    meta: dict[str, Any] = Field(default_factory=dict)
    error: ApiError | None = None


class LoadStatusUpdateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    data: dict[str, Any]
    meta: dict[str, Any] = Field(default_factory=dict)
    error: ApiError | None = None
